libsolidity/codegen/ExpressionCompiler.cpp

```diff
@@ -1282,35 +1299,53 @@ bool ExpressionCompiler::visit(FunctionCall const& _functionCall)
 
 			if (hasSelectorOrSignature)
 			{
-				// stack: <selector> <memory pointer>
-				solAssert(arguments.size() >= 1, "");
-				Type const* selectorType = arguments[0]->annotation().type;
-				utils().moveIntoStack(selectorType->sizeOnStack());
-				Type const* dataOnStack = selectorType;
-				// stack: <memory pointer> <selector>
-				if (function.kind() == FunctionType::Kind::ABIEncodeWithSignature)
+				Type const* dataOnStack = nullptr;
+
+				if (function.kind() == FunctionType::Kind::ABIEncodeCall)
 				{
-					// hash the signature
-					if (auto const* stringType = dynamic_cast<StringLiteralType const*>(selectorType))
-					{
-						m_context << util::selectorFromSignature(stringType->value());
-						dataOnStack = TypeProvider::fixedBytes(4);
-					}
+					auto const functionPtr = dynamic_cast<FunctionTypePointer>(arguments[0]->annotation().type);
+					solAssert(functionPtr, "");
+
+					if (functionPtr->hasDeclaration())
+						m_context << util::selectorFromSignature(functionPtr->externalSignature());
 					else
 					{
-						utils().fetchFreeMemoryPointer();
-						// stack: <memory pointer> <selector> <free mem ptr>
-						utils().packedEncode(TypePointers{selectorType}, TypePointers());
-						utils().toSizeAfterFreeMemoryPointer();
-						m_context << Instruction::KECCAK256;
-						// stack: <memory pointer> <hash>
-
-						dataOnStack = TypeProvider::fixedBytes(32);
+						m_context << Instruction::SWAP1 << Instruction::POP;
+						/// need to store it as bytes4
+						utils().leftShiftNumberOnStack(224);
```

Comment:
Instead of shifting, you could set the `dataOnStack` to `uint32` and the conversion below will do the rest.